src/box/xlog.c

```diff
@@ -710,16 +718,25 @@ xdir_collect_garbage(struct xdir *dir, int64_t signature, unsigned flags)
 	struct vclock *vclock;
 	while ((vclock = vclockset_first(&dir->index)) != NULL &&
 	       vclock_sum(vclock) < signature) {
+		int64_t del_signature = vclock_sum(vclock);
 		const char *filename =
-			xdir_format_filename(dir, vclock_sum(vclock), NONE);
+			xdir_format_filename(dir, del_signature, NONE);
 		xlog_remove_file(filename, rm_flags);
+		if (xlog_remove_callback != NULL)
+			xlog_remove_callback(dir->dirname, del_signature);
```

Comment:
1. The callback should follow the flags, i.e. called asynchronously if `XLOG_RM_ASYNC` is set.
2. If the callback fails, the corresponding xlog file shouldn't be deleted.
3. `xdir_collect_garbage_cb` is an unfortunate name because it looks like it's a callback. Let's just add a new argument to `xdir_collect_garbage` - it's used in just a few places so I think it's acceptable.
4. Please do this in a separate commit.

Reply:
Done. Now the callback and reader & writer constructors consume the `.snap` file name instead of the memtx directory path.